File: Next-Up-Hardwood-Hoops-main/courthoops/simulation/engine/shot_engine.py
from typing import Callable, Optional, Dict

from courthoops.domain.game.events import Event
from courthoops.domain.game.possession import Possession
from courthoops.infra.utils.rng import IRNG
from courthoops.simulation.playbyplay.event_builder import EventBuilder


class ShotEngine:
    """Resolves shot outcomes with probability based on lineup strength."""

    def __init__(
        self,
        rng: IRNG,
        offense_strength_fn: Callable[[str], float],
        defense_strength_fn: Callable[[str], float],
        player_lookup: Optional[Dict[str, object]] = None,
        event_builder: Optional[EventBuilder] = None,
    ):
        self.rng = rng
        self.offense_strength_fn = offense_strength_fn
        self.defense_strength_fn = defense_strength_fn
        self.player_lookup = player_lookup or {}
        self.event_builder = event_builder or EventBuilder()

    def resolve_shot(self, possession: Possession, shooter_id: str | None = None, assisted_by: str | None = None) -> Event:
        offense_strength = self.offense_strength_fn(possession.offense_team_id)
        defense_strength = self.defense_strength_fn(possession.defense_team_id)

        # Bias shot type by play_type: spot_up -> more threes, post_up -> fewer threes
        if possession.play_type == "spot_up":
            three_bias = 0.55
        elif possession.play_type == "post_up":
            three_bias = 0.1
        else:  # pnr or default
            three_bias = 0.35

        is_three = self.rng.random() < three_bias
        is_free_throw = False
        drew_and_one = False

        base_prob = 0.45 + (offense_strength - defense_strength) / 200.0
        if is_three:
            base_prob -= 0.05
        if possession.play_type == "post_up":
            base_prob += 0.02  # slight boost for close looks
        elif possession.play_type == "pnr":
            base_prob += 0.01
        base_prob += self._shooter_bonus(shooter_id, possession.play_type, is_three)
        # Attribute influence placeholders; caller can override strength_fn to include player-level detail
        make_prob = min(max(base_prob, 0.20), 0.8)
        made = self.rng.random() < make_prob
        points = 3 if is_three else 2

        # Occasional shooting fouls generating free throws
        if not made and self.rng.random() < 0.08:
            is_free_throw = True
            points = 1
            is_three = False

        event_type = "shot_made" if made else "shot_missed"
        desc = f"{possession.offense_team_id} {'drains' if made else 'misses'} a {points}-pt attempt."
        payload = {
            "team_id": possession.offense_team_id,
            "player_id": shooter_id,
            "assist_player_id": assisted_by if made else None,
            "points": points if made else 0,
            "period": possession.period,
            "is_three": is_three,
            "play_type": possession.play_type,
            "is_free_throw": is_free_throw,
            "and_one": drew_and_one,
        }
        return self.event_builder.build(event_type, desc, payload)

    def _shooter_bonus(self, shooter_id: Optional[str], play_type: Optional[str], is_three: bool) -> float:
        player = self.player_lookup.get(shooter_id) if shooter_id else None
        if not player:
            return 0.0
        attrs = getattr(player, "attributes", None)
        if not attrs:
            return 0.0
        bonus = 0.0
        if is_three:
            bonus += (attrs.three_point - 70) / 300.0
        elif play_type == "post_up":
            bonus += (attrs.inside - 70) / 300.0 + (attrs.dunk - 70) / 400.0
        else:  # pnr/spot midrange/inside
            bonus += (attrs.mid_range - 70) / 300.0 + (attrs.layup - 70) / 400.0
        return bonus

File: Scripts/Core/Simulation/Engines/ShotEngine.cs
using System;
using System.Collections.Generic;
using HardwoodHoops.Core.Domain.Games;
using HardwoodHoops.Core.Domain.Players;
using HardwoodHoops.Core.Simulation.PlayByPlay;
using HardwoodHoops.Core.Simulation.Rng;

namespace HardwoodHoops.Core.Simulation.Engines;

public sealed class ShotEngine
{
    private readonly IRng _rng;
    private readonly Func<string, double> _offenseStrength;
    private readonly Func<string, double> _defenseStrength;
    private readonly Dictionary<string, Player> _playerLookup;
    private readonly EventBuilder _eventBuilder;

    public ShotEngine(
        IRng rng,
        Func<string, double> offenseStrength,
        Func<string, double> defenseStrength,
        Dictionary<string, Player>? playerLookup = null,
        EventBuilder? eventBuilder = null)
    {
        _rng = rng;
        _offenseStrength = offenseStrength;
        _defenseStrength = defenseStrength;
        _playerLookup = playerLookup ?? new Dictionary<string, Player>();
        _eventBuilder = eventBuilder ?? new EventBuilder();
    }

    public GameEvent ResolveShot(string offenseTeamId, string defenseTeamId, string? shooterId, string? assistedBy, string playType)
    {
        var offenseStrength = _offenseStrength(offenseTeamId);
        var defenseStrength = _defenseStrength(defenseTeamId);

        var threeBias = playType switch
        {
            "spot_up" => 0.55,
            "post_up" => 0.1,
            _ => 0.35
        };

        var isThree = _rng.NextDouble() < threeBias;
        var baseProb = 0.45 + (offenseStrength - defenseStrength) / 200.0;
        if (isThree)
        {
            baseProb -= 0.05;
        }

        if (playType == "post_up")
        {
            baseProb += 0.02;
        }
        else if (playType == "pnr")
        {
            baseProb += 0.01;
        }

        baseProb += ShooterBonus(shooterId, playType, isThree);
        var makeProb = Math.Clamp(baseProb, 0.20, 0.8);
        var made = _rng.NextDouble() < makeProb;
        var points = isThree ? 3 : 2;
        var isFreeThrow = false;
        var andOne = false;

        if (!made && _rng.NextDouble() < 0.08)
        {
            isFreeThrow = true;
            points = 1;
            isThree = false;
        }

        var eventType = made ? "shot_made" : "shot_missed";
        var description = $"{offenseTeamId} {(made ? "drains" : "misses")} a {points}-pt attempt.";
        var payload = new Dictionary<string, object>
        {
            ["team_id"] = offenseTeamId,
            ["player_id"] = shooterId ?? string.Empty,
            ["assist_player_id"] = made ? assistedBy ?? string.Empty : string.Empty,
            ["points"] = made ? points : 0,
            ["is_three"] = isThree,
            ["play_type"] = playType,
            ["is_free_throw"] = isFreeThrow,
            ["and_one"] = andOne
        };

        return _eventBuilder.Build(eventType, description, payload);
    }

    private double ShooterBonus(string? shooterId, string playType, bool isThree)
    {
        if (shooterId is null || !_playerLookup.TryGetValue(shooterId, out var player))
        {
            return 0.0;
        }

        var attrs = player.Attributes;
        if (isThree)
        {
            return (attrs.ThreePoint - 70) / 300.0;
        }

        if (playType == "post_up")
        {
            return (attrs.Inside - 70) / 300.0 + (attrs.Dunk - 70) / 400.0;
        }

        return (attrs.MidRange - 70) / 300.0 + (attrs.Layup - 70) / 400.0;
    }
}
